### Коломієць Микола Ом-4, Варіант 8

### Варіант 8

8.1. Економіка країни розбита на дві виробничі галузі (промисловість та сільське господарство). За минулий рік
повний випуск промислових виробництв у вартісній формі був розподілений таким чином:
- 650 млн. грн. для виробничих потреб промисловості;
- 370 млн. грн. для виробничих потреб сільського господарства;
- 750 млн. грн. для споживання населення (згідно попиту на цю продукцію).

В той же час повний випуск сільськогосподарської продукції (у вартісній формі) був розподілений таким чином:
- 300 млн. грн. для виробничих потреб промисловості;
- 350 млн. грн. для виробничих потреб сільського господарства;
- 550 млн. грн. для споживання населення (згідно попиту на цю продукцію).

Розрахувати ціни на промислову та сільськогосподарську продукцію, якщо відомо, що додана вартість в цінах
складає: 

- 0.5 - для промисловості; 
- 0.7 - для сільського господарства.

In [2]:
import matplotlib.pyplot as plt
import sympy as sp
import numpy as np
import scipy as scp

In [8]:
x11, x12, x21, x22, y1, y2 = sp.symbols('x_{11} x_{12} x_{21} x_{22} y_1 y_2')
x1 = x11 + x12 + y1
x2 = x21 + x22 + y2

In [9]:
params = {
    # промисловість
    x11: 650,
    x12: 370,
    y1: 750,
    # сільське господарство
    x21: 300,
    x22: 350,
    y2: 550
}

In [10]:
A = sp.Matrix([[x11/x1, x12/x2], [x21/x1, x22/x2]])
A

Matrix([
[x_{11}/(x_{11} + x_{12} + y_1), x_{12}/(x_{21} + x_{22} + y_2)],
[x_{21}/(x_{11} + x_{12} + y_1), x_{22}/(x_{21} + x_{22} + y_2)]])

In [12]:
A = A.subs(params)
A

Matrix([
[65/177, 37/120],
[ 10/59,   7/24]])

In [17]:
s = sp.Matrix([0.5, 0.7]).T
s

Matrix([[0.5, 0.7]])

In [19]:
E = sp.Matrix([[1, 0], [0, 1]])
B = sp.Inverse(E - A)
p = s * B
p

Matrix([[1.19411414982164, 1.50802615933413]])

Висновок ціна на промислову продукцію складає p1 = 1.19 мільйонів гривень, на сільськогосподарську продукцію p2 = 1.51 мільйонів гривень

8.2. Знайти власні числа матриці А, коефіцієнти характеристичного поліному, її число Фробеніуса, правий та лівий

вектори Фробеніуса. Зробити висновок про продуктивність даної матриці: А $\begin{pmatrix}0.45 & 0.3 & 0.2 \\ 0.2 & 0.25 & 0.4 \\ 0.3 & 0.35 & 0.3\end{pmatrix}$. 

Для цієї матриці знайти матрицю повних витрат В. Дослідити на збіжність суму ряду $Е + А + А^2 +... + А^N$ до матриці повних витрат
(критерій збіжності - величини елементів відповідних матриць відрізняються менше, ніж на 0.01).

Знайти вектор кінцевого випуску, якщо вектор кінцевого споживання продукції $у = \begin{pmatrix} 100 \\ 70 \\ 110\end{pmatrix}$.

In [31]:
A = sp.Matrix([[0.45, 0.3, 0.2], [0.2, 0.25, 0.4], [0.3, 0.35, 0.3]])
# Знайдемо власні числа
eigenvalues = A.eigenvals()
for eigenvalue in eigenvalues.keys():
    print(eigenvalue)

0.917307682853362
0.164080842588588
-0.0813885254419498


In [32]:
# знайдемо коефіцієнти характеристичного поліному матриці A
A.charpoly().coeffs()

[1.00000000000000, -1.00000000000000, 0.0625000000000000, 0.0122500000000000]

In [46]:
# знайдемо число Форбеніуса
forbenius = max(eigenvalues.keys())
print(forbenius)
# знайдемо власний вектор Форбеніуса
eigenvector = A.eigenvects()[0][2][0]
eigenvector

0.917307682853362


Matrix([
[0.598814750700913],
[0.536116387953526],
[0.594979086111997]])

In [48]:
forbenius < 1 # матриця пародуктивна

True

In [55]:
E = sp.Matrix([[1, 0, 0], [0, 1, 0], [0, 0, 1]])
B = sp.Inverse(E - A) * E
B

Matrix([
[ 5.1505016722408, 3.74581939799331, 3.61204013377926],
[3.47826086956522, 4.34782608695652, 3.47826086956522],
[3.94648829431438, 3.77926421404682, 4.71571906354515]])

In [62]:
A_sum = E
k = 0
Ak = A
while abs((A_sum - B)[2]) > 0.01:
    A_sum += Ak
    Ak *= A
    k+=1
print(k)

69


In [63]:
B * sp.Matrix([100, 70, 110]) # вектор кінцевого випуску продукції

Matrix([
[1174.58193979933],
[1034.78260869565],
[1177.92642140468]])